File: app/src/main/java/com/barakamntmkrib/MainActivity.java
package com.barakamntmkrib;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;

import android.annotation.SuppressLint;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.ContextMenu;
import android.view.MotionEvent;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import java.io.*;
import java.net.*;
import java.util.Enumeration;

public class MainActivity extends AppCompatActivity {

    private TextView txt_url;
    private EditText txt_ip;
    private EditText txt_port;
    private EditText txt_link;


    private String host;
    private int port;


    private Socket socket;
    private PrintWriter printwriter;
    private boolean connection;

    public static final String CONNECTION_PREFS = "connection";
    private SharedPreferences pref;
    private SharedPreferences.Editor editor;

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES); //For night mode theme
        setContentView(R.layout.activity_main);


        txt_url = findViewById(R.id.txt_url);
        registerForContextMenu(txt_url);
        txt_ip = findViewById(R.id.txt_ip);
        txt_port = findViewById(R.id.txt_port);
        txt_link = findViewById(R.id.txt_link);
        //The listener of a drawableEnd button for clear a TextInputEditText
        txt_link.setOnTouchListener(new View.OnTouchListener() {
            @SuppressLint("ClickableViewAccessibility")
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if(event.getAction() == MotionEvent.ACTION_UP) {
                    final EditText textView = (EditText)v;
                    if(event.getX() >= textView.getWidth() - textView.getCompoundPaddingEnd()) {
                        textView.setText(""); //Clear a view, example: EditText or TextView
                        textView.requestFocus();
                        InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
                        imm.showSoftInput(textView, InputMethodManager.SHOW_IMPLICIT);
                        return true;
                    }
                }
                return false;
            }
        });
        pref = getSharedPreferences(CONNECTION_PREFS, MODE_PRIVATE);
        if (pref.contains("ip_address") && pref.contains("port_number")) {
            load_preferences();
        }

        // Get intent, action and MIME type
        Intent intent = getIntent();
        String action = intent.getAction();
        String type = intent.getType();

        if (Intent.ACTION_SEND.equals(action) && type != null) {
            handleSendText(intent); // Handle text being sent
            /*if ("text/plain".equals(type)) {

            }*/
        }


    }
    public void send_command(View view){
        String command;
        switch (view.getId()) {
            case (R.id.volume_reduce_button):
                command = "volume down";
                break;
            case (R.id.volume_increase_button):
                command = "volume up";
                break;
            case (R.id.rewind_button):
                command = "left arrow";
                break;
            case (R.id.play_pause_button):
                command = "space";
                break;
            case (R.id.forward_button):
                command = "right arrow";
                break;
            case (R.id.subtitle_button):
                command = "s";
                break;
            case (R.id.stop_button):
                command = ";";
                break;
            case (R.id.next_chapter_button):
                command = "page down";
                break;
            case (R.id.previous_chapter_button):
                command = "page up";
                break;
            case (R.id.skip_forward_button):
                command = "ctrl+page down";
                break;
            case (R.id.skip_back_button):
                command = "ctrl+page up";
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + view.getId());
        }
        host = txt_ip.getText().toString();
        port = Integer.parseInt(txt_port.getText().toString());
        open_link(command, host, port);
    }

    void handleSendText(Intent intent) {
        host = txt_ip.getText().toString();
        port = Integer.parseInt(txt_port.getText().toString());
        String sharedText = intent.getStringExtra(Intent.EXTRA_TEXT);
        if (sharedText != null) {
            // Update UI to reflect text being shared
            txt_url.setText(sharedText);
            open_link(sharedText, host, port);
        } else {
            String error = "Error getting the link";
            txt_url.setText(error);
        }
    }

    public String get_ip_address(){
        StringBuilder ip_address = new StringBuilder();
        try {
            for (Enumeration<NetworkInterface> network_enum = NetworkInterface.getNetworkInterfaces(); network_enum.hasMoreElements();) {
                NetworkInterface network_interface = network_enum.nextElement();
                for (Enumeration<InetAddress> ip_address_enum = network_interface.getInetAddresses(); ip_address_enum.hasMoreElements();) {
                    InetAddress inetAddress = ip_address_enum.nextElement();
                    if (!inetAddress.isLoopbackAddress() && !inetAddress.isLinkLocalAddress() && inetAddress.isSiteLocalAddress()) {
                        ip_address.append(inetAddress.getHostAddress().toString()).append("\n");
                        Toast.makeText(getApplicationContext(), ip_address.toString(), Toast.LENGTH_SHORT).show();
                    }

                }
            }
        } catch (SocketException ex) {
            Log.e("LOG_TAG", ex.toString());
        }
        return ip_address.toString();
    }

    public void save_preferences(View view) {
        /*editor = getSharedPreferences(CONNECTION_PREFS, MODE_PRIVATE).edit();
        editor.putString("ip_address", txt_ip.getText().toString());
        editor.putInt("port_number", Integer.parseInt(txt_port.getText().toString()));
        editor.apply();*/
        UdpListener udp_listener = new UdpListener(this);
        Thread udp_client = new Thread(udp_listener);
        udp_client.start();
        // txt_ip.setText(get_ip_address());
    }

    public void load_preferences() {
        String ip = pref.getString("ip_address", "192.168.1.106");//"No ip defined" is the default value.
        int port = pref.getInt("port_number", 1007); //1007 is the default port.
        txt_ip.setText(ip);
        txt_port.setText(String.valueOf(port));
    }

    public void open_external_link(View view) {
        host = txt_ip.getText().toString();
        port = Integer.parseInt(txt_port.getText().toString());
        String url = txt_link.getText().toString();
        txt_url.setText(url);
        open_link(url, host, port);
    }
    @Override
    public void onCreateContextMenu(ContextMenu menu, View v,
                                    ContextMenu.ContextMenuInfo menuInfo) {
        // user has long pressed your TextView
        menu.add(0, v.getId(), 0,
                "Copy");

        // cast the received View to TextView so that you can get its text
        TextView yourTextView = (TextView) v;

        // place your TextView's text in clipboard
        ClipboardManager clipboard = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);

        ClipData clipData = android.content.ClipData
                .newPlainText("last link", yourTextView.getText());
        clipboard.setPrimaryClip(clipData);
    }

    public void open_link(final String command, final String host, final int port) {
        new Thread(new Runnable() {

            @Override
            public void run() {
                try {

                    //System.out.println("Your current Hostname : " + host);
                    socket = new Socket(host, port);


                    printwriter = new PrintWriter(socket.getOutputStream(), true);
                    printwriter.write(command); // write the message to output stream
                    printwriter.flush();
                    printwriter.close();

                    Log.d("socket", "connected");


                    // Toast in background because Toast cannot be in main thread you have to create runOnuithread.
                    // this is run on ui thread where dialogs and all other GUI will run.
                    if (socket.isConnected()) {
                        MainActivity.this.runOnUiThread(new Runnable() {
                            public void run() {
                                Toast.makeText(getApplicationContext(), "Command sent", Toast.LENGTH_SHORT).show();
                            }
                        });
                    }
                } catch (UnknownHostException e2) {
                    MainActivity.this.runOnUiThread(new Runnable() {
                        public void run() {
                            //Do your UI operations like dialog opening or Toast here
                            Toast.makeText(getApplicationContext(), "Unknown host please make sure IP address", Toast.LENGTH_SHORT).show();

                        }
                    });

                } catch (IOException e1) {
                    Log.d("socket", "IOException");
                    MainActivity.this.runOnUiThread(new Runnable() {
                        public void run() {
                            //Do your UI operations like dialog opening or Toast here
                            Toast.makeText(getApplicationContext(), "Error Occurred ", Toast.LENGTH_SHORT).show();
                        }
                    });
                }

            }
        }).start();
    }


}

File: app/src/main/java/com/barakamntmkrib/MainActivity.kt
package com.barakamntmkrib

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.content.*
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.util.Log
import android.view.ContextMenu
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.Switch
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import com.barakamntmkrib.UdpListener.TextFromRunnable
import java.io.IOException
import java.io.PrintWriter
import java.net.Socket
import java.net.UnknownHostException

class MainActivity : AppCompatActivity() {
    private var txtServerStatus: TextView? = null
    private var txtIp: EditText? = null
    private var txtLink: EditText? = null
    private var switchBluetooth: Switch? = null

    private var socket: Socket? = null
    private var host: String? = null
    private val port = 1007
    //to implement: detect if the server is connected or not,
    // also timeout udp listener
    private var connection = false

    // This is the activity main thread Handler.
    private var updateUIHandler: Handler? = null
    private val bluetoothOn = "bluetooth_on"
    private val bluetoothOff = "bluetooth_off"

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES) //For night mode theme
        setContentView(R.layout.activity_main)

        // Initialize Handler.
        createUpdateUiHandler()
        //
        getServerAddress()

        // Get link from the share menu
        val intent = intent
        val action = intent.action
        val type = intent.type
        if (Intent.ACTION_SEND == action && type != null) {
            sharedTextHandler(intent) // Handle text being sent
        }

        // Register for broadcasts on BluetoothAdapter state change
        val filter = IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
        registerReceiver(mReceiver, filter)

        //
        txtServerStatus = findViewById(R.id.txt_server_status)
        // Override onCreateContextMenu to change behavior
        registerForContextMenu(txtServerStatus)

        txtIp = findViewById(R.id.txt_ip)
        switchBluetooth = findViewById(R.id.bluetooth_switch)

        txtLink = findViewById(R.id.txt_link)
        //calling the listener of a drawableEnd button to clear a TextInputEditText
        txtLink?.setOnTouchListener(clearTextInputDrawableOnTouchListener())



        switchBluetooth?.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                // The toggle is enabled
                sendCommand(bluetoothOn)
            } else {
                // The toggle is disabled
                sendCommand(bluetoothOff)
            }
        }
    }

    private fun clearTextInputDrawableOnTouchListener(): OnTouchListener? {
        //The listener of a drawableEnd button to clear a TextInputEditText
        return OnTouchListener { v, event ->
            if (event.action == MotionEvent.ACTION_UP) {
                val textView = v as EditText
                if (event.x >= textView.width - textView.compoundPaddingEnd) {
                    textView.setText("") //Clear a view, example: EditText or TextView
                    textView.requestFocus()
                    val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                    imm.showSoftInput(textView, InputMethodManager.SHOW_IMPLICIT)
                    return@OnTouchListener true
                }
            }
            false
        }
    }

    @SuppressLint("HandlerLeak")
    private fun createUpdateUiHandler() {
        if (updateUIHandler == null) {
            updateUIHandler = object : Handler() {
                override fun handleMessage(msg: Message) {
                    // Means the message is sent from child thread.
                    if (msg.what == MESSAGE_UPDATE_TEXT_CHILD_THREAD) {
                        // Update ui in main thread.
                        updateIpValue(msg.obj.toString())
                    }
                }
            }
        }
    }

    fun getCommand(view: View) {
        val command: String = when (view.id) {
            R.id.volume_reduce_button -> "volume down"
            R.id.volume_increase_button -> "volume up"
            R.id.rewind_button -> "left arrow"
            R.id.play_pause_button -> "space"
            R.id.forward_button -> "right arrow"
            R.id.subtitle_button -> "s"
            R.id.audio_button -> "a"
            R.id.stop_button -> ";"
            R.id.next_chapter_button -> "page down"
            R.id.previous_chapter_button -> "page up"
            R.id.skip_forward_button -> "ctrl+page down"
            R.id.skip_back_button -> "ctrl+page up"
            R.id.btn_turn_off_screen -> "turn_off_screen"
            else -> throw IllegalStateException("Unexpected value: " + view.id)
        }
        sendCommand(command)
    }

    fun sendCommand(command: String) {
        host = txtIp?.text.toString()
        openLink(command, host)
    }

    private fun sharedTextHandler(intent: Intent) {
        host = txtIp!!.text.toString()
        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
        if (sharedText != null) {
            // Update UI to reflect text being shared
            txtServerStatus!!.text = sharedText
            openLink(sharedText, host)
        } else {
            val error = "Error getting the link"
            txtServerStatus!!.text = error
        }
    }

    fun updateIpValue(ipAddress: String?) {
        txtIp!!.setText(ipAddress)
    }

    @Throws(InterruptedException::class)
    fun getServerAddress() {
        // udp
        val udpListener = UdpListener(object : TextFromRunnable {
            override fun sendString(text: String?) {
                // Build message object.
                val message = Message()
                // Set message type.
                message.what = MESSAGE_UPDATE_TEXT_CHILD_THREAD
                message.obj = text
                // Send message to main thread Handler.
                updateUIHandler!!.sendMessage(message)
            }
        })
        val udpClient = Thread(udpListener)
        udpClient.start()
        udpClient.join()
    }


    fun openExternalLink(v: View) {
        host = txtIp!!.text.toString()
        val url = txtLink!!.text.toString()
        txtServerStatus!!.text = url
        openLink(url, host)
    }



    private fun openLink(command: String, host: String?) {
        Thread(Runnable {
            try {
                var printwriter: PrintWriter? = null
                //System.out.println("Your current Hostname : " + host);
                socket = Socket(host, port)
                printwriter = PrintWriter(socket!!.getOutputStream(), true)
                printwriter.write(command) // write the message to output stream
                printwriter.flush()
                printwriter.close()
                Log.d("socket", "connected")


                // Toast in background because Toast cannot be in main thread you have to create runOnuithread.
                // this is run on ui thread where dialogs and all other GUI will run.
                if (socket!!.isConnected) {
                    runOnUiThread { Toast.makeText(applicationContext, "Command sent", Toast.LENGTH_SHORT).show() }
                }
            } catch (e2: UnknownHostException) {
                runOnUiThread { //Do your UI operations like dialog opening or Toast here
                    Toast.makeText(applicationContext, "Unknown host please make sure IP address", Toast.LENGTH_SHORT).show()
                }
            } catch (e1: IOException) {
                Log.d("socket", "IOException")
                runOnUiThread { //Do your UI operations like dialog opening or Toast here
                    Toast.makeText(applicationContext, "Error Occurred ", Toast.LENGTH_SHORT).show()
                }
            }
        }).start()
    }

    private val mReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action!!
            if (action == BluetoothAdapter.ACTION_STATE_CHANGED) {
                val state = intent.getIntExtra(BluetoothAdapter.EXTRA_STATE,
                        BluetoothAdapter.ERROR)
                when (state) {
                    BluetoothAdapter.STATE_OFF -> bluetoothStateChange("STATE_OFF")
                    BluetoothAdapter.STATE_TURNING_OFF -> bluetoothStateChange("STATE_TURNING_OFF")
                    BluetoothAdapter.STATE_ON -> bluetoothStateChange("STATE_ON")
                    BluetoothAdapter.STATE_TURNING_ON -> bluetoothStateChange("STATE_TURNING_ON")
                }
            }
        }

        private fun bluetoothStateChange(bluetooth_state: String) {
            if (bluetooth_state == "STATE_TURNING_OFF") {
                // when the state is off turn bluetooth on on computer
                sendCommand(bluetoothOn)

            } else if (bluetooth_state == "STATE_TURNING_ON") {
                // to disconnect the bluetooth on the computer
                sendCommand(bluetoothOff)

            }
        }
    }
    override fun onCreateContextMenu(menu: ContextMenu, view: View,
                                     menuInfo: ContextMenu.ContextMenuInfo) {
        // user has long pressed your TextView
        menu.add(0, view.id, 0,
                "Copy")

        // cast the received View to TextView so that you can get its text
        val yourTextView = view as TextView

        // place your TextView's text in clipboard
        val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clipData = ClipData
                .newPlainText("last link", yourTextView.text)
        clipboard.primaryClip = clipData
    }

    public override fun onDestroy() {
        super.onDestroy()

        /* ... */

        // Unregister broadcast listeners
        unregisterReceiver(mReceiver)
    }

    companion object {
        // Message type code.
        private const val MESSAGE_UPDATE_TEXT_CHILD_THREAD = 1
    }
}